test/plugin/base_test.rb

```diff
@@ -0,0 +1,1318 @@
+                     "Test with #{n} logs failed. \n#{entry}"
+      end
+    end
+  end
+
+  def test_cloudfunctions_logs_text_not_matched
+    setup_gce_metadata_stubs
+    setup_cloudfunctions_metadata_stubs
+    [1, 2, 3, 5, 11, 50].each do |n|
+      @logs_sent = []
+      setup_logging_stubs do
+        d = create_driver(APPLICATION_DEFAULT_CONFIG, CLOUDFUNCTIONS_TAG)
+        # The test driver doesn't clear its buffer of entries after running, so
+        # do it manually here.
+        d.instance_variable_get('@entries').clear
+        n.times { |i| d.emit(cloudfunctions_log_entry_text_not_matched(i)) }
+        d.run
+      end
+      verify_log_entries(
+        n, CLOUDFUNCTIONS_TEXT_NOT_MATCHED_PARAMS) do |entry|
+        assert_equal 'INFO', entry['metadata']['severity'],
+                     "Test with #{n} logs failed. \n#{entry}"
+      end
+    end
+  end
+
+  def test_multiple_cloudfunctions_logs_tag_not_matched
+    setup_gce_metadata_stubs
+    setup_cloudfunctions_metadata_stubs
+    [1, 2, 3, 5, 11, 50].each do |n|
+      @logs_sent = []
+      setup_logging_stubs do
+        d = create_driver(APPLICATION_DEFAULT_CONFIG, CONTAINER_TAG)
+        # The test driver doesn't clear its buffer of entries after running, so
+        # do it manually here.
+        d.instance_variable_get('@entries').clear
+        n.times { |i| d.emit(cloudfunctions_log_entry(i)) }
+        d.run
+      end
+      i = 0
+      verify_log_entries(n, CONTAINER_FROM_TAG_PARAMS, '') do |entry|
+        assert_equal '[D][2015-09-25T12:34:56.789Z][123-0] test log entry ' \
+                     "#{i}", entry['textPayload'],
+                     "Test with #{n} logs failed. \n#{entry}"
+        i += 1
+      end
+    end
+  end
+
+  def test_http_request_from_record
+    setup_gce_metadata_stubs
+    setup_logging_stubs do
+      d = create_driver
+      d.emit('httpRequest' => http_request_message)
+      d.run
+    end
+    verify_log_entries(1, COMPUTE_PARAMS, 'httpRequest') do |entry|
+      assert_equal http_request_message, entry['httpRequest'], entry
+      assert_nil get_fields(entry['structPayload'])['httpRequest'], entry
+    end
+  end
+
+  def test_http_request_partial_from_record
+    setup_gce_metadata_stubs
+    setup_logging_stubs do
+      d = create_driver
+      d.emit('httpRequest' => http_request_message.merge(
+        'otherKey' => 'value'))
+      d.run
+    end
+    verify_log_entries(1, COMPUTE_PARAMS, 'httpRequest') do |entry|
+      assert_equal http_request_message, entry['httpRequest'], entry
+      fields = get_fields(entry['structPayload'])
+      request = get_fields(get_struct(fields['httpRequest']))
+      assert_equal 'value', get_string(request['otherKey']), entry
+    end
+  end
+
+  def test_http_request_without_referer_from_record
+    setup_gce_metadata_stubs
+    setup_logging_stubs do
+      d = create_driver
+      d.emit('httpRequest' => http_request_message_without_referer)
```

Comment:
A comment as been added. For the uppercase concern, similar to the other comment, Ruby does not allow not defining a constant in the shared module. Per discussion offline, this will remain as a method (actually a wrapper around a constant).